File: packages/tolk/bench/runtime/bench_runtime.py
#!/usr/bin/env python3
"""Compare execution-only throughput with an explicit tinygrad checkout.

Run either companion first, using the same DEV for both:
  python packages/tolk/bench/runtime/bench_runtime.py OUT --reference PATH_TO_REFERENCE

Both sides use realized zero-filled inputs, two JIT warmup/capture calls and
adaptive replay timing. Compilation and input-buffer allocation are outside
the timed region. Repeat alternating runs on a quiet host before drawing conclusions.
"""

import argparse
import hashlib
import json
import math
import os
import subprocess
import sys
from pathlib import Path
import time

parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("out_dir", nargs="?", default=".")
parser.add_argument("--reference", type=Path, required=True)
args = parser.parse_args()
reference = args.reference.resolve()
reference_package = (reference / "tinygrad").resolve()
if not (reference_package / "__init__.py").is_file():
    parser.error("--reference must contain the tinygrad Python package")
sys.path.insert(0, str(reference))

# Match the OCaml runner: CPU by default, or the explicitly selected backend.
os.environ.setdefault("DEV", "CPU")
os.environ.setdefault("NO_COLOR", "1")

import tinygrad  # noqa: E402

if Path(tinygrad.__file__).resolve().parent != reference_package:
    raise RuntimeError("imported tinygrad does not belong to --reference")

from tinygrad import Device, Tensor, TinyJit  # noqa: E402
from tinygrad.device import Buffer  # noqa: E402
from tinygrad.dtype import dtypes  # noqa: E402
from tinygrad.helpers import GlobalCounters  # noqa: E402

BACKEND = Device.DEFAULT
F32_BYTES = 4
BUF_ELEMS = 16 * 1024 * 1024
TARGET_S = 1.5
MIN_K = 5
MAX_K = 5000


def sync():
    Device[BACKEND].synchronize()


def time_replay(call):
    """Adaptive execution-only timing: one estimate call sizes K, then median
    and min per-replay nanoseconds over K samples."""
    t0 = time.perf_counter_ns()
    call()
    est_s = (time.perf_counter_ns() - t0) / 1e9
    k = int(max(1.0, TARGET_S / max(est_s, 1e-9)))
    k = max(MIN_K, min(MAX_K, k))
    samples = []
    storage_before = GlobalCounters.mem_used
    for _ in range(k):
        t0 = time.perf_counter_ns()
        call()
        samples.append(float(time.perf_counter_ns() - t0))
    storage_after = GlobalCounters.mem_used
    samples.sort()
    return samples[k // 2], samples[0], k, storage_before, storage_after


def time_compute(build, inputs):
    jf = TinyJit(build)

    def call():
        jf(*inputs)
        sync()

    # Warm and capture; the estimate call below is the first replay.
    call()
    call()
    return time_replay(call)


def input_tensor(shape):
    return Tensor(bytes(math.prod(shape) * F32_BYTES), dtype=dtypes.float32,
                  device=BACKEND).reshape(shape).realize()


def matmul_bench(n):
    a = input_tensor((n, n))
    b = input_tensor((n, n))
    median, minimum, k, storage_before, storage_after = time_compute(lambda a, b: (a @ b).realize(), (a, b))
    flops = 2.0 * n * n * n
    return {"bench": "matmul", "size": str(n), "unit": "GFLOP/s",
            "amount": flops, "median_ns": median, "min_ns": minimum, "k": k,
            "storage_bytes_before": storage_before, "storage_bytes_after": storage_after}


def elementwise_bench(n):
    a = input_tensor((n,))
    b = input_tensor((n,))
    c = input_tensor((n,))
    median, minimum, k, storage_before, storage_after = time_compute(
        lambda a, b, c: (a + b * c).realize(), (a, b, c))
    return {"bench": "elementwise", "size": "16M", "unit": "GB/s",
            "amount": 4.0 * n * F32_BYTES,
            "median_ns": median, "min_ns": minimum, "k": k,
            "storage_bytes_before": storage_before, "storage_bytes_after": storage_after}


def reduce_bench(n):
    x = input_tensor((n,))
    median, minimum, k, storage_before, storage_after = time_compute(lambda x: x.sum().realize(), (x,))
    return {"bench": "reduce", "size": "16M", "unit": "GB/s",
            "amount": float(n) * F32_BYTES,
            "median_ns": median, "min_ns": minimum, "k": k,
            "storage_bytes_before": storage_before, "storage_bytes_after": storage_after}


def copy_bench(n):
    dev = Device[BACKEND]
    buf = Buffer(BACKEND, n, dtypes.float32).allocate()
    host = memoryview(bytearray(n * F32_BYTES))

    def call():
        buf.host[:] = host
        dev.synchronize()

    call()
    median, minimum, k, storage_before, storage_after = time_replay(call)
    return {"bench": "copy", "size": "16M", "unit": "GB/s",
            "amount": float(n) * F32_BYTES,
            "median_ns": median, "min_ns": minimum, "k": k,
            "storage_bytes_before": storage_before, "storage_bytes_after": storage_after}


BENCHES = [
    lambda: matmul_bench(512),
    lambda: matmul_bench(1024),
    lambda: elementwise_bench(BUF_ELEMS),
    lambda: reduce_bench(BUF_ELEMS),
    lambda: copy_bench(BUF_ELEMS),
]


def per_ns(amount, ns):
    # One flop per ns is a GFLOP/s; one byte per ns is a GB/s.
    return amount / ns if ns > 0 else 0.0


def run_reference():
    rows = {}
    for bench in BENCHES:
        row = bench()
        rows[(row["bench"], row["size"])] = row
    return rows


def main():
    out_dir = args.out_dir

    tolk_file = Path(out_dir, "tolk_runtime.json")
    tolk_rows = json.loads(tolk_file.read_text()) if tolk_file.exists() else []

    if any(row["backend"] != BACKEND for row in tolk_rows):
        raise ValueError("both runtime benchmarks must use the same DEV backend")
    revision = (subprocess.check_output(
        ["git", "-C", str(reference), "rev-parse", "HEAD"], text=True).strip()
        if (reference / ".git").exists() else None)
    digest = hashlib.sha256()
    for source in sorted((reference / "tinygrad").rglob("*.py")):
        digest.update(str(source.relative_to(reference)).encode() + b"\0")
        digest.update(source.read_bytes() + b"\0")
    ref = run_reference()
    provenance = {"reference": str(reference), "imported_package": str(reference_package),
                  "revision": revision,
                  "python_source_sha256": digest.hexdigest(), "python": sys.version,
                  "target": str(Device[BACKEND].renderer.target),
                  "rows": list(ref.values())}
    Path(out_dir, "tinygrad_runtime.json").write_text(json.dumps(provenance, indent=2) + "\n")

    if not tolk_rows:
        print(f"wrote {len(ref)} reference rows for {BACKEND} to {out_dir}")
        return

    header = ["bench", "size", "backend", "tolk", "tg", "unit"]
    table = []
    for t in tolk_rows:
        key = (t["bench"], t["size"])
        r = ref.get(key)
        tg = per_ns(r["amount"], r["median_ns"]) if r else None
        table.append({
            "bench": t["bench"], "size": t["size"], "backend": t["backend"],
            "tolk": t["median"], "tolk_peak": t["peak"],
            "tg": tg,
            "tg_peak": per_ns(r["amount"], r["min_ns"]) if r else None,
            "unit": t["unit"],
        })

    print(f"{BACKEND} execution-only throughput against tinygrad {revision or digest.hexdigest()}")
    rows = [header] + [[
        r["bench"], r["size"], r["backend"], f"{r['tolk']:.2f}",
        f"{r['tg']:.2f}" if r["tg"] is not None else "-", r["unit"],
    ] for r in table]
    widths = [max(len(row[i]) for row in rows) for i in range(len(header))]
    for i, row in enumerate(rows):
        print("  ".join(c.ljust(widths[j]) for j, c in enumerate(row)))
        if i == 0:
            print("  ".join("-" * widths[j] for j in range(len(header))))

    tsv_path = os.path.join(out_dir, "runtime.tsv")
    cols = ["bench", "size", "backend", "tolk_median", "tolk_peak",
            "tg_median", "tg_peak", "unit"]
    with open(tsv_path, "w") as f:
        f.write("\t".join(cols) + "\n")
        for r in table:
            f.write("\t".join([
                r["bench"], r["size"], r["backend"],
                f"{r['tolk']:.6f}", f"{r['tolk_peak']:.6f}",
                f"{r['tg']:.6f}" if r["tg"] is not None else "",
                f"{r['tg_peak']:.6f}" if r["tg_peak"] is not None else "",
                r["unit"],
            ]) + "\n")
    print(f"\nwrote {tsv_path}")


if __name__ == "__main__":
    main()

File: packages/tolk/bench/runtime/bench_runtime.ml
(*---------------------------------------------------------------------------
  Copyright (c) 2026 The Raven authors. All rights reserved.
  SPDX-License-Identifier: ISC
  ---------------------------------------------------------------------------*)

(* Indicative runtime-throughput benchmarks. Each workload is compiled once
   through the capture-and-replay JIT (or, for the copy case, allocated once)
   and then replayed K times; the timed region is execution only, with compile
   and scheduling held outside the loop.

   The four workloads and their throughput units:
     matmul       a @ b, float32          GFLOP/s   (2*M*N*K flops)
     elementwise  a + b * c over a buffer  GB/s      (3 reads + 1 write)
     reduce       sum over a buffer        GB/s      (1 read)
     copy         host -> device buffer    GB/s      (bytes uploaded)

   Run the Python companion against an explicit tinygrad checkout for a
   like-for-like CPU comparison. Repeat alternating runs on a quiet host.
   The backend defaults to CPU unless [DEV] selects another device; its
   name is recorded in every row. Allocation deltas cover the calling domain
   and include timing-loop overhead; storage bytes count live logical buffers,
   not native allocator reservations or backend object handles. *)

module Dtype = Tolk_uop.Dtype
module Dev = Tolk.Device
module El = Tolk_frontend.Elementwise
module Rd = Tolk_frontend.Reduce
module Op = Tolk_frontend.Op
module Run = Tolk_frontend.Run
module Jit = Tolk_frontend.Jit

let keep x = ignore (Sys.opaque_identity x)
let now () = Thumper_clock.elapsed_ns ()

let f32_bytes = 4

(* A zero-filled float32 tensor of [shape]. Values are immaterial to timing. *)
let input shape =
  let n = List.fold_left ( * ) 1 shape in
  Run.of_bytes ~dtype:Dtype.float32 ~shape (Bytes.make (n * f32_bytes) '\000')

(* Adaptive replay timing. One estimate call sizes K so the timed run spans
   roughly [target_s]; K is clamped so tiny and huge workloads both behave.
   Returns median and minimum per-replay nanoseconds and the K used. *)
let target_s = 1.5
let min_k = 5
let max_k = 5000

type timing = {
  median_ns : float;
  min_ns : float;
  k : int;
  allocated_words : float;
  promoted_words : float;
  major_words : float;
  storage_bytes_before : int;
  storage_bytes_after : int;
}

let time_replay call =
  let est =
    let t0 = now () in
    call ();
    Int64.to_float (Int64.sub (now ()) t0) /. 1e9
  in
  let k =
    int_of_float (Float.max 1. (target_s /. Float.max est 1e-9))
    |> max min_k |> min max_k
  in
  let samples = Array.make k 0. in
  let storage_bytes_before = Tolk.Helpers.Global_counters.mem_used () in
  let minor, promoted, major = Gc.counters () in
  for i = 0 to k - 1 do
    let t0 = now () in
    call ();
    samples.(i) <- Int64.to_float (Int64.sub (now ()) t0)
  done;
  let minor_after, promoted_after, major_after = Gc.counters () in
  let storage_bytes_after = Tolk.Helpers.Global_counters.mem_used () in
  Array.sort Float.compare samples;
  let per_call words = words /. float_of_int k in
  {
    median_ns = samples.(k / 2); min_ns = samples.(0); k;
    allocated_words = per_call
        ((minor_after -. minor) +. (major_after -. major) -. (promoted_after -. promoted));
    promoted_words = per_call (promoted_after -. promoted);
    major_words = per_call (major_after -. major);
    storage_bytes_before; storage_bytes_after;
  }

(* A replayable compute workload: build the graph from realized inputs, warm
   and capture through the JIT, then time execution-only replays. *)
let time_compute ~build inputs =
  let jit = Jit.create ~outputs:(fun tensor -> [tensor]) (fun ins ~vars:_ -> Run.realize (build ins)) in
  Array.iter (fun t -> keep (Run.realize t)) inputs;
  let dev = Run.device () in
  let call () =
    keep (Jit.call jit inputs);
    Dev.synchronize dev
  in
  call ();
  call ();
  time_replay call

(* Host -> device upload: allocate the buffer and host bytes once, then time
   the copyin primitive. *)
let time_copy n =
  let dev = Run.device () in
  let buf = Dev.create_buffer ~size:n ~dtype:Dtype.float32 dev in
  Dev.Buffer.ensure_allocated buf;
  let host = Bytes.make (n * f32_bytes) '\000' in
  let call () =
    Dev.Buffer.copyin buf host;
    Dev.synchronize dev
  in
  call ();
  time_replay call

(* Rows *)

type row = {
  bench : string;
  size : string;
  unit_ : string;
  amount : float;  (* flops or bytes moved per replay *)
  timing : timing;
}

(* Throughput is [amount /. ns]: one flop per ns is a GFLOP/s, one byte per ns
   is a GB/s, so both units share the same conversion. *)
let per_ns amount ns = amount /. ns

let matmul_row n =
  let a = input [ n; n ] and b = input [ n; n ] in
  let timing =
    time_compute ~build:(fun ins -> Op.matmul ins.(0) ins.(1)) [| a; b |]
  in
  let flops = 2.0 *. float_of_int n *. float_of_int n *. float_of_int n in
  {
    bench = "matmul";
    size = string_of_int n;
    unit_ = "GFLOP/s";
    amount = flops;
    timing;
  }

let elementwise_row n =
  let a = input [ n ] and b = input [ n ] and c = input [ n ] in
  let timing =
    time_compute
      ~build:(fun ins -> El.add ins.(0) (El.mul ins.(1) ins.(2)))
      [| a; b; c |]
  in
  let bytes = 4.0 *. float_of_int n *. float_of_int f32_bytes in
  {
    bench = "elementwise";
    size = "16M";
    unit_ = "GB/s";
    amount = bytes;
    timing;
  }

let reduce_row n =
  let x = input [ n ] in
  let timing =
    time_compute ~build:(fun ins -> Rd.sum ins.(0)) [| x |]
  in
  let bytes = float_of_int n *. float_of_int f32_bytes in
  {
    bench = "reduce";
    size = "16M";
    unit_ = "GB/s";
    amount = bytes;
    timing;
  }

let copy_row n =
  let timing = time_copy n in
  let bytes = float_of_int n *. float_of_int f32_bytes in
  {
    bench = "copy";
    size = "16M";
    unit_ = "GB/s";
    amount = bytes;
    timing;
  }

(* Output *)

let write_file path contents =
  let oc = open_out path in
  output_string oc contents;
  close_out oc

let row_json backend r =
  Printf.sprintf
    {|{"bench":"%s","size":"%s","backend":"%s","unit":"%s","median":%.6f,"peak":%.6f,"k":%d,"allocated_words_per_replay":%.0f,"promoted_words_per_replay":%.0f,"major_words_per_replay":%.0f,"storage_bytes_before":%d,"storage_bytes_after":%d}|}
    r.bench r.size backend r.unit_ (per_ns r.amount r.timing.median_ns)
    (per_ns r.amount r.timing.min_ns) r.timing.k
    r.timing.allocated_words r.timing.promoted_words r.timing.major_words
    r.timing.storage_bytes_before r.timing.storage_bytes_after

let print_table backend rows =
  Printf.printf "\ntolk runtime throughput (backend %s)\n" backend;
  Printf.printf "%-12s  %-6s  %10s  %10s  %8s  %6s\n" "bench" "size" "median"
    "peak" "unit" "k";
  List.iter
    (fun r ->
      Printf.printf "%-12s  %-6s  %10.2f  %10.2f  %8s  %6d\n" r.bench r.size
        (per_ns r.amount r.timing.median_ns)
        (per_ns r.amount r.timing.min_ns)
        r.unit_ r.timing.k)
    rows

let () =
  if Sys.getenv_opt "DEV" = None then Unix.putenv "DEV" "CPU";
  let out_dir = if Array.length Sys.argv > 1 then Sys.argv.(1) else "." in
  let rows =
    [
      matmul_row 512;
      matmul_row 1024;
      elementwise_row (16 * 1024 * 1024);
      reduce_row (16 * 1024 * 1024);
      copy_row (16 * 1024 * 1024);
    ]
  in
  let backend = Dev.name (Run.device ()) in
  print_table backend rows;
  let json =
    "[\n  "
    ^ String.concat ",\n  " (List.map (row_json backend) rows)
    ^ "\n]\n"
  in
  write_file (Filename.concat out_dir "tolk_runtime.json") json;
  Printf.printf "\nwrote %d rows to %s/tolk_runtime.json\n" (List.length rows)
    out_dir
